elastic/logstash
# Add system properties to configure Jackson's stream read constraints

## Release notes


Added `jvm.options` properties to configure the Jackson read constraints defaults (Maximum Number value length, Maximum String value length, and Maximum Nesting depth).

## What does this PR do?



This PR adds 3 new custom system properties into the `jvm.options`. Those new properties are used to fine-tune the default Jackson stream read constraints values, which are used by Jackson to guard against malicious input by preventing processing of too big inputs. The added properties can be found [here](https://github.com/elastic/logstash/pull/15720/files#diff-d592447cb7aefcc8d1929fd64fa5dcb40a77e476e221f9690a88e01a002f264bR74-R88).

The intention of early set the Jackson defaults during the startup (`runner.rb`) was to ensure that all Logstash's functionalities that rely on Jackson are using those settings values.

## Why is it important/What is the impact to the user?



Although the Jackson's defaults works for the majority of users, it might be too restrictive - and a blocker - for pipelines receiving big payload requests or with a deeper level of nested objects. 

Having the possibility of tuning those properties is essential to unblock users with that use-case and for a more accurate guard against malicious inputs.

## Checklist




## How to test this PR locally



- Set any property value to a lower/higher value, e.g.: `-Dlogstash.jackson.stream-read-constraints.max-string-length=2`
- Configure and run a pipeline with a `beats` input
- Send any data to this pipeline that exceeds the configured property value
- Check the logs for: `com.fasterxml.jackson.core.exc.StreamConstraintsException: String length (24) exceeds the maximum length (2)`

## Related issues


- Closes https://github.com/elastic/ingest-dev/issues/2638
